ping/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+mod protocol_builder;
+mod protocol_generated;
+
+use crate::protocol_generated::p2p::ping::*;
+use flatbuffers::{get_root, FlatBufferBuilder};
+use fnv::FnvHashMap;
+use futures::sync::mpsc::Sender;
+use log::debug;
+use p2p::{
+    context::{ServiceContext, SessionContext},
+    traits::{ProtocolMeta, ServiceProtocol},
+    ProtocolId, SessionId,
+};
+use std::{
+    str,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tokio::codec::length_delimited::LengthDelimitedCodec;
+
+const SEND_PING_TOKEN: u64 = 0;
+const CHECK_TIMEOUT_TOKEN: u64 = 1;
+// TODO: replace this with real PeerId
+type PeerId = SessionId;
```

Comment:
`peerid` just a hash for public key，I think this is a very easy thing to implement, even in the upper layer agreement. But I am not sure what Hash algorithm to use.

Reply:
In our context, we need a hash function that security enough(avoid collision between different public keys) and should be as fast as possible.

https://dadario.com.br/cryptographic-and-non-cryptographic-hash-functions/
https://github.com/oconnor663/blake2b_simd

IMO blake2b is a good choice, it has the same security level with sha3, and benchmark shows it also has high performance. (In our scenario we need to hash a public key(65 bytes) and the performance result may be different, we need to do some benchmark to make sure).

I think we cloud use blake2b, we can replace it anytime before main-net launch.